lib/src/scanner.dart

```diff
@@ -2,4 +2,219 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-class Scanner {}
+import 'dart:collection';
+
+import 'package:charcode/ascii.dart';
+import 'package:string_scanner/string_scanner.dart';
+
+class Scanner extends IterableBase<Token> implements Iterable<Token> {
+  Scanner(this.sourceUrl, this.dx);
+
+  final Uri sourceUrl;
+  final String dx;
+
+  @override
+  Iterator<Token> get iterator => new _ScannerIterator(sourceUrl, dx);
+}
+
+/// Produces a stream of [Token]s from dx source code.
+class _ScannerIterator implements Iterator<Token> {
+  _ScannerIterator(Uri sourceUrl, String dx)
+      : _scanner = new SpanScanner(dx, sourceUrl: sourceUrl);
+
+  final SpanScanner _scanner;
+
+  @override
+  Token get current => _current;
+  Token _current;
+
+  @override
+  bool moveNext() {
+    _consumeWhitespace();
+
+    if (_scanner.isDone) {
+      return false;
+    }
+
+    if (!_scanNext()) {
+      _scanner.expectDone();
+    }
+
+    return true;
+  }
+
+  bool _scanNext() =>
+    _scanKeyword() ||
+    _scanIdentifier() ||
+    _scanSyntax();
+
+  bool _scanKeyword() {
+    if (_scanner.scan(_Patterns.keyword)) {
+      _current = Keyword.lookupByName(_lastString);
+      return true;
+    }
+    return false;
+  }
+
+  bool _scanIdentifier() {
+    if (_scanner.scan(_Patterns.identifier)) {
+      _current = new Identifier(_lastString);
+      return true;
+    }
+    return false;
+  }
+
+  bool _scanSyntax() {
+    if (_scanner.scan(_Patterns.syntax)) {
+      _current = Syntax.lookupByName(_lastString);
+      return true;
+    }
+    return false;
+  }
+
+  void _consumeWhitespace() {
+    while (!_scanner.isDone && isWhitespace(_scanner.peekChar())) {
+      _scanner.readChar();
+    }
+  }
+
+  String get _lastString => _scanner.lastMatch.group(0);
```

Comment:
Nit: I believe `group` is just an alias for `[]`.

Reply:
Done.